src/meta/typesmap.go

```diff
@@ -72,30 +86,32 @@ func (m TypesMap) IsResolved() bool {
 
 // NewEmptyTypesMap creates new type map that has no types in it
 func NewEmptyTypesMap(cap int) TypesMap {
-	return TypesMap{m: make(map[string]struct{}, cap)}
+	return TypesMap{m: make(map[Type]struct{}, cap)}
 }
 
-func NewTypesMapFromTypes(types []Type) TypesMap {
-	m := make(map[string]struct{}, len(types))
-	for _, typ := range types {
-		s := typ.Elem
-		for i := 0; i < typ.Dims; i++ {
-			s = WrapArrayOf(s)
-		}
-		m[s] = struct{}{}
+func NewTypesMapFromPhpDocTypes(phpDocTypes []PhpDocType) TypesMap {
+	m := make(RawTypesMap, len(phpDocTypes))
+
+	for _, phpDocType := range phpDocTypes {
+		typ := NewTypeFromPhpDocType(phpDocType)
+		m.Add(typ)
 	}
+
 	return TypesMap{m: m}
 }
 
 // NewTypesMap returns new TypesMap that is initialized with the provided types (separated by "|" symbol)
 func NewTypesMap(str string) TypesMap {
-	m := make(map[string]struct{}, strings.Count(str, "|")+1)
-	for _, s := range strings.Split(str, "|") {
-		if IsArrayType(s) {
-			s = WrapArrayOf(strings.TrimSuffix(s, "[]"))
+	m := make(RawTypesMap, strings.Count(str, "|")+1)
+
+	for _, typeStr := range strings.Split(str, "|") {
+		typ := NewType(typeStr)
+		if typ.IsArray() {
+			typ = WrapArrayOf(typ.ElementType())
 		}
-		m[s] = struct{}{}
+		m.Add(typ)
```

Comment:
I don't think it is really obvious when you are adding something to a value type (it is RawTypesMap, it is not *RawTypesMap) does this mean that the value will change? That is the reason why TypesMap provides `Append()` functions that return you (potentially a new) value.